### PR TITLE
Add _gels functions

### LINKED ISSUE #7316: dgels is missing
``dgels`` doesn't seem to be exposed in ``sp.linalg.lapack``. Would it make sense to expose it? Also it could be added as a driver to ``lstsq``. All the current drivers are SVD based, so it would be nice to have a method based on the QR decomposition. It should be a bit faster too. I can work on a PR if this sounds reasonable.

### PR DESCRIPTION
**Pull request by thrasibule**
This solve the least square problem using the qr decomposition which is faster than gelsy, gelss and gelsd which are svd based. This closes #7316 .

### CODE REVIEW

File: scipy/linalg/tests/test_lapack.py [hunk 1]
```diff
@@ -138,6 +138,66 @@ def test_clapack(self):
...
+                                                (a1, b1))
+
+            m, n = a1.shape
+            if len(b1.shape) == 2:
+                nrhs = b1.shape[1]
+            else:
+                nrhs = 1
+
+            # Request of sizes
+            work, info = gels_lwork(m,n,nrhs)
```
Comment (ev-br) [inline scipy/linalg/tests/test_lapack.py]: PEP8 says spaces after commas please. Also several other PEP8 violations, https://travis-ci.org/scipy/scipy/jobs/227275928

File: scipy/linalg/tests/test_lapack.py [hunk 2]
```diff
@@ -138,6 +138,66 @@ def test_clapack(self):
...
+
+            lqr, x, info = gels(a1, b1, lwork)
+            assert_allclose(x[:-1], np.array([-14.333333333333323,
+                                              14.999999999999991], dtype=dtype),
+                            rtol=25*np.finfo(dtype).eps)
+            assert_allclose(lqr, np.array([[-8.1240384 , -9.6011363 ],
+                                           [ 0.43840236,  0.90453403],
+                                           [ 0.76720414,  0.90907633]],
+                                          dtype=dtype),
+                            rtol=25*np.finfo(dtype).eps)
```
Comment (ev-br) [inline scipy/linalg/tests/test_lapack.py]: https://travis-ci.org/scipy/scipy/jobs/227275929 looks real. On my machine the abs difference is of the order of 1e-9.

File: scipy/linalg/flapack.pyf.src [hunk 3]
```diff
@@ -1188,6 +1188,98 @@ interface
...
+     integer depend(b),intent(hide):: nrhs = shape(b,1)
+     integer depend(a),intent(hide):: lda = shape(a,0)
+     integer intent(hide),depend(a):: m = shape(a,0)
+     integer intent(hide),depend(a):: n = shape(a,1)
+     integer depend(b),intent(hide):: ldb = shape(b,0)
+
+     <ftype2> dimension(m,nrhs),intent(in,out,copy,out=x) :: b
+     <ftype2> dimension(m,n),intent(in,out,copy,out=lqr) :: a
+     <ftype2> depend(lwork),dimension(MAX(lwork,1)),intent(hide,cache):: work
+     integer intent(out)::info
```
Comment (pv) [inline scipy/linalg/flapack.pyf.src]: There's no declaration for `lwork` here?
Maybe it works without, but I'd be explicit here.

File: scipy/linalg/tests/test_lapack.py (hunk 1 above)
Comment (pv) [inline scipy/linalg/tests/test_lapack.py]: Minor comment: I'd maybe rewrite this to check for `lwork = _compute_lwork(gels_lwork, m, n, nrhs)` where that routine comes from `from scipy.linalg.lapack import _compute_lwork`

File: scipy/linalg/flapack.pyf.src [hunk 4]
```diff
@@ -1188,6 +1188,98 @@ interface
...
+     fortranname <prefix2c>gels
+     callstatement (*f2py_func)(trans,&m,&n,&nrhs,a,&lda,b,&ldb,work,&lwork,&info)
+     callprotoargument char*,int*,int*,int*,<ctype2c>*,int*,<ctype2c>*,int*,<ctype2c>*,int*,int*
+
+     character optional,intent(in),check(*trans=='N'||*trans=='T'):: trans = 'N'
+     integer depend(b),intent(hide):: nrhs = shape(b,1)
+     integer depend(a),intent(hide):: lda = shape(a,0)
+     integer intent(hide),depend(a):: m = shape(a,0)
+     integer intent(hide),depend(a):: n = shape(a,1)
+     integer depend(b),intent(hide):: ldb = shape(b,0)
```
Comment (pv) [inline scipy/linalg/flapack.pyf.src]: Missing declaration for lwork?

**Review by pv — COMMENTED** [review]
LGTM mainly, although I'd make sure lwork is declared in the pyf file

**Pull request merged**